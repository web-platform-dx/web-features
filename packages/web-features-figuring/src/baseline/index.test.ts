import assert from "node:assert/strict";

import * as chai from "chai";
import chaiJestSnapshot from "chai-jest-snapshot";

import { computeBaseline } from ".";

chai.use(chaiJestSnapshot);

describe("computeBaseline", function () {
  before(function () {
    chaiJestSnapshot.resetSnapshotRegistry();
  });

  beforeEach(function () {
    chaiJestSnapshot.configureUsingMochaContext(this);
  });

  it("returns something sensible for the most complex features", function () {
    // These are some of the most "complex" BCD features, given approximately by:
    // const complexity = JSON.stringify(data.__compat, undefined, 2).split("\n").length
    const result = Object.fromEntries(
      [
        "api.Document.visibilitychange_event",
        "css.types.basic-shape.path",
        "api.DOMMatrix.DOMMatrix",
      ].map((key) => [
        key,
        computeBaseline({ compatKeys: [key], checkAncestors: false }),
      ]),
    );
    assert.equal(
      result["api.Document.visibilitychange_event"]?.baseline,
      "high",
    );
    assert.equal(result["css.types.basic-shape.path"]?.baseline, false);
    assert.equal(result["api.DOMMatrix.DOMMatrix"]?.baseline, "high");
    chai.expect(result).to.matchSnapshot();
  });

  it("returns something sensible for the least complex features", function () {
    // These are some of the least "complex" BCD features
    const result = Object.fromEntries(
      [
        "css.properties.counter-reset.reset_does_not_affect_siblings",
        "html.global_attributes.exportparts",
        "html.elements.form.target",
      ].map((key) => [
        key,
        computeBaseline({ compatKeys: [key], checkAncestors: false }),
      ]),
    );

    assert.equal(
      result["css.properties.counter-reset.reset_does_not_affect_siblings"]
        ?.baseline,
      false,
    );
    assert.equal(
      result["html.global_attributes.exportparts"]?.baseline,
      "high",
    );
    assert.equal(result["html.elements.form.target"]?.baseline, "high");
    chai.expect(result).to.matchSnapshot();
  });

  it("returns a result for a feature", function () {
    const result = computeBaseline({
      compatKeys: ["css.properties.border-color"],
      checkAncestors: false,
    });
    assert.equal(result.baseline, "high");
    assert.equal(result.baseline_low_date, "2015-07-28"); // The first release of Edge, the youngest release in consideration
    assert.equal(result.baseline_high_date, "2018-01-28"); // 30 months later
  });
});
