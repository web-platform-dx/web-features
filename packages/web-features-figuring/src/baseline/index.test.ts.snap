// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`computeBaseline returns something sensible for the least complex features 1`] = `
Object {
  "css.properties.counter-reset.reset_does_not_affect_siblings": "{
  \\"baseline\\": false,
  \\"support\\": {
    \\"firefox\\": \\"82\\",
    \\"firefox_android\\": \\"82\\"
  }
}",
  "html.elements.form.target": "{
  \\"baseline\\": \\"high\\",
  \\"baseline_low_date\\": \\"2015-07-28\\",
  \\"baseline_high_date\\": \\"2018-01-28\\",
  \\"support\\": {
    \\"chrome\\": \\"1\\",
    \\"chrome_android\\": \\"18\\",
    \\"edge\\": \\"12\\",
    \\"firefox\\": \\"1\\",
    \\"firefox_android\\": \\"4\\",
    \\"safari\\": \\"3\\",
    \\"safari_ios\\": \\"2\\"
  }
}",
  "html.global_attributes.exportparts": "{
  \\"baseline\\": \\"high\\",
  \\"baseline_low_date\\": \\"2020-07-28\\",
  \\"baseline_high_date\\": \\"2023-01-28\\",
  \\"support\\": {
    \\"chrome\\": \\"73\\",
    \\"chrome_android\\": \\"73\\",
    \\"edge\\": \\"79\\",
    \\"firefox\\": \\"72\\",
    \\"firefox_android\\": \\"79\\",
    \\"safari\\": \\"13.1\\",
    \\"safari_ios\\": \\"13.4\\"
  }
}",
}
`;

exports[`computeBaseline returns something sensible for the most complex features 1`] = `
Object {
  "api.DOMMatrix.DOMMatrix": "{
  \\"baseline\\": \\"high\\",
  \\"baseline_low_date\\": \\"2020-01-15\\",
  \\"baseline_high_date\\": \\"2022-07-15\\",
  \\"support\\": {
    \\"chrome\\": \\"61\\",
    \\"chrome_android\\": \\"61\\",
    \\"edge\\": \\"79\\",
    \\"firefox\\": \\"33\\",
    \\"firefox_android\\": \\"33\\",
    \\"safari\\": \\"11\\",
    \\"safari_ios\\": \\"11\\"
  }
}",
  "api.Document.visibilitychange_event": "{
  \\"baseline\\": \\"high\\",
  \\"baseline_low_date\\": \\"2021-04-26\\",
  \\"baseline_high_date\\": \\"2023-10-26\\",
  \\"support\\": {
    \\"chrome\\": \\"62\\",
    \\"chrome_android\\": \\"62\\",
    \\"edge\\": \\"18\\",
    \\"firefox\\": \\"56\\",
    \\"firefox_android\\": \\"56\\",
    \\"safari\\": \\"14.1\\",
    \\"safari_ios\\": \\"14.5\\"
  }
}",
  "css.types.basic-shape.path": "{
  \\"baseline\\": false,
  \\"support\\": {}
}",
}
`;
